presto-main/src/main/java/com/facebook/presto/server/SerializedPageWriteListener.java

```diff
@@ -0,0 +1,102 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.server;
+
+import com.facebook.airlift.log.Logger;
+import com.facebook.presto.execution.buffer.SerializedPage;
+import io.airlift.slice.Slice;
+
+import javax.servlet.AsyncContext;
+import javax.servlet.ServletOutputStream;
+import javax.servlet.WriteListener;
+
+import java.io.IOException;
+import java.util.Queue;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.airlift.slice.SizeOf.SIZE_OF_BYTE;
+import static io.airlift.slice.SizeOf.SIZE_OF_INT;
+import static io.airlift.slice.Slices.allocate;
+import static java.util.Objects.requireNonNull;
+import static sun.misc.Unsafe.ARRAY_BYTE_BASE_OFFSET;
+
+public class SerializedPageWriteListener
+        implements WriteListener
+{
+    private final Logger log = Logger.get(SerializedPageWriteListener.class);
+
+    private static final int METADATA_BUFFER_SIZE = 16;
+    private final Queue<SerializedPage> serializedPages;
+    private final AsyncContext asyncContext;
+    private final ServletOutputStream output;
+    private final Slice slice;
+    private SerializedPage page;
+
+    public SerializedPageWriteListener(Queue<SerializedPage> serializedPages,
+            AsyncContext asyncContext,
+            ServletOutputStream output)
+    {
+        this.serializedPages = requireNonNull(serializedPages, "serializedPages is null");
+        this.asyncContext = requireNonNull(asyncContext, "asyncContext is null");
+        this.output = requireNonNull(output, "output is null");
+        this.slice = allocate(METADATA_BUFFER_SIZE);
+    }
+
+    @Override
+    public void onWritePossible()
+            throws IOException
+    {
+        while (output.isReady()) {
+            if (writeComplete()) {
+                asyncContext.complete();
+                return;
+            }
+
+            if (page == null) {
+                page = serializedPages.poll();
+
+                int bufferPosition = 0;
+
+                slice.setInt(bufferPosition, page.getPositionCount());
+                bufferPosition += SIZE_OF_INT;
+                slice.setByte(bufferPosition, page.getPageCodecMarkers());
+                bufferPosition += SIZE_OF_BYTE;
+                slice.setInt(bufferPosition, page.getUncompressedSizeInBytes());
+                bufferPosition += SIZE_OF_INT;
+                slice.setInt(bufferPosition, page.getSizeInBytes());
+                bufferPosition += SIZE_OF_INT;
+
+                output.write(slice.byteArray(), 0, bufferPosition);
+            }
+            else {
+                Object base = page.getSlice().getBase();
+                checkArgument(base instanceof byte[], "serialization type only supports byte[]");
+                output.write((byte[]) base, (int) ((page.getSlice().getAddress() - ARRAY_BYTE_BASE_OFFSET)), page.getSizeInBytes());
+                page = null;
```

Comment:
I would recommend either wrapping the `OutputStream` here with `OutputStreamSliceOutput` and calling `writeBytes`, or calling the `getBytes` method on Slice itself (which encapsulates the logic of whether or not this has an underlying byte array).

Reply:
`getBytes` seems to make a copy

Reply:
I mean the overload `public void getBytes(int index, OutputStream out, int length)`

This won't make a copy unless necessary

Reply:
@tdcmeehan Do you mean extending the `Slice` interface?

Reply:
I mean this public method: https://github.com/prestodb/slice/blob/master/src/main/java/com/facebook/slice/Slice.java#L650

Or `OutputStreamSliceOutput`

Reply:
Oh, right. I didn't know such a method exist in the Slice interface. I thought all methods there take SliceOutput / Input. That makes perfect sense. Thanks for the suggestion.